File: clients/ios/Lextures/Core/LMS/ContentToolPack1Logic.swift
import Foundation

/// Pure CT.M5 pack-1 decisions — attempt/reveal gating, commit immutability,
/// Class Pulse poll backoff, offline action queue ordering, client allowlist,
/// and Review/SRS cache invalidation keys. No networking.
enum ContentToolPack1Logic {
    static let pack1ToolIds: Set<String> = [
        "inline_questions",
        "predict_reveal",
        "class_pulse",
        "flashcards",
    ]

    /// Per-tool client allowlist (rollout). Empty entry removes a renderer without a release.
    static var clientAllowlist: Set<String> = pack1ToolIds

    static let classPulsePollIntervalMs = 30_000
    static let classPulseMaxBackoffMs = 120_000
    static let flashcardRatings = ["again", "hard", "good", "easy"]

    struct PendingAction: Equatable {
        var instanceId: String
        var toolId: String
        var action: String
        var sequence: Int64
        var payloadJSON: String
    }

    // MARK: - Allowlist / registry

    static func isClientAllowlisted(
        _ toolId: String,
        allowlist: Set<String> = clientAllowlist
    ) -> Bool {
        allowlist.contains(toolId)
    }

    static func allowlistedToolIds(allowlist: Set<String> = clientAllowlist) -> Set<String> {
        pack1ToolIds.intersection(allowlist)
    }

    static func conflictPolicy(for toolId: String) -> ContentToolHostLogic.ConflictPolicy {
        if toolId == "flashcards" { return .merge }
        return .serverWins
    }

    // MARK: - Offline action queue (tool-local; CT.M3 host never queues actions)

    /// Pack-1 exceptions: `submit` and `rate` may queue; vote/commit/aggregate never do.
    static func canQueueActionOffline(toolId: String, action: String) -> Bool {
        switch (toolId, action) {
        case ("inline_questions", "submit"), ("flashcards", "rate"):
            return true
        default:
            return false
        }
    }

    static func orderPendingActions(_ items: [PendingAction]) -> [PendingAction] {
        items.sorted {
            if $0.instanceId != $1.instanceId { return $0.instanceId < $1.instanceId }
            return $0.sequence < $1.sequence
        }
    }

    // MARK: - inline_questions

    static func parseAttemptsConfig(_ raw: JSONValue?) -> Int? {
        // nil = unlimited
        guard let raw else { return 2 }
        switch raw {
        case .string(let text) where text.lowercased() == "unlimited":
            return nil
        case .string(let text):
            return Int(text).map { max(1, $0) } ?? 2
        case .number(let number):
            return max(1, Int(number.rounded()))
        default:
            return 2
        }
    }

    static func attemptsUsed(answers: [String: JSONValue], questionId: String) -> Int {
        guard case .object(let question)? = answers[questionId],
              case .array(let attempts)? = question["attempts"]
        else { return 0 }
        return attempts.count
    }

    static func canSubmit(
        answers: [String: JSONValue],
        questionId: String,
        maxAttempts: Int?,
        readOnly: Bool
    ) -> Bool {
        if readOnly { return false }
        guard let maxAttempts else { return true }
        return attemptsUsed(answers: answers, questionId: questionId) < maxAttempts
    }

    static func isSequentiallyUnlocked(
        questions: [String],
        answers: [String: JSONValue],
        questionId: String,
        sequential: Bool
    ) -> Bool {
        guard sequential else { return true }
        for qid in questions {
            if qid == questionId { return true }
            if attemptsUsed(answers: answers, questionId: qid) == 0 { return false }
        }
        return false
    }

    static func shuffleStable<T>(_ items: [T], seed: String, id: (T) -> String) -> [T] {
        guard items.count > 1 else { return items }
        var out = items
        var h: UInt32 = 0
        for byte in seed.utf8 {
            h = h &* 31 &+ UInt32(byte)
        }
        if out.count > 1 {
            for i in stride(from: out.count - 1, through: 1, by: -1) {
                h = h &* 1_103_515_245 &+ 12_345
                let j = Int(h % UInt32(i + 1))
                out.swapAt(i, j)
            }
        }
        _ = id
        return out
    }

    // MARK: - predict_reveal

    static func isCommitted(_ state: JSONValue?) -> Bool {
        guard let committed = ContentToolHostLogic.stringField(state, key: "committedAt"),
              !committed.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else { return false }
        return true
    }

    /// Reveal is gated on server-confirmed commit — never on local draft alone.
    static func canShowReveal(committed: Bool, hasRevealPayload: Bool) -> Bool {
        committed && hasRevealPayload
    }

    static func canEditPrediction(committed: Bool, readOnly: Bool) -> Bool {
        !committed && !readOnly
    }

    // MARK: - class_pulse

    static func hasVoted(votes: [JSONValue], round: Int = 1) -> Bool {
        votes.contains { vote in
            guard case .object(let obj) = vote else { return false }
            if case .number(let roundNumber) = obj["round"] { return Int(roundNumber) == round }
            if case .string(let roundText) = obj["round"], let parsed = Int(roundText) {
                return parsed == round
            }
            return false
        }
    }

    static func shouldPollAggregate(visible: Bool, hasVoted: Bool) -> Bool {
        visible && hasVoted
    }

    /// Visibility-gated poll with linear backoff on failures (caps at 120s).
    static func nextPollDelayMs(consecutiveFailures: Int, baseMs: Int = classPulsePollIntervalMs) -> Int {
        if consecutiveFailures <= 0 { return baseMs }
        let factor = 1 << min(consecutiveFailures, 3)
        return min(classPulseMaxBackoffMs, baseMs * factor)
    }

    // MARK: - flashcards / Review reconciliation

    static func isValidRating(_ rating: String) -> Bool {
        flashcardRatings.contains(rating.lowercased())
    }

    /// Same SRS queue as Review — invalidate these keys after a session so due/streak refresh.
    static func reviewCacheKeysToInvalidate() -> [String] {
        [OfflineCacheKey.reviewQueue(), OfflineCacheKey.reviewStats()]
    }

    static func shouldDoubleCountReviewSubmit(toolId: String) -> Bool {
        // Never call POST /learners/.../review from the flashcards tool.
        toolId != "flashcards"
    }

    // MARK: - Unknown-field preservation

    /// Apply known keys without dropping unknown server fields (NFR backward compat).
    static func mergePreservingUnknown(
        base: [String: JSONValue],
        patch: [String: JSONValue]
    ) -> [String: JSONValue] {
        var out = base
        for (k, v) in patch { out[k] = v }
        return out
    }

    // MARK: - JSON helpers used by renderers

    static func objectMap(_ value: JSONValue?) -> [String: JSONValue] {
        guard case .object(let obj) = value else { return [:] }
        return obj
    }

    static func arrayField(_ value: JSONValue?, key: String) -> [JSONValue] {
        guard case .object(let obj) = value, case .array(let arr) = obj[key] else { return [] }
        return arr
    }

    static func boolField(_ value: JSONValue?, key: String) -> Bool? {
        guard case .object(let obj) = value, let field = obj[key] else { return nil }
        if case .bool(let flag) = field { return flag }
        return nil
    }

    static func numberField(_ value: JSONValue?, key: String) -> Double? {
        guard case .object(let obj) = value, let field = obj[key] else { return nil }
        if case .number(let number) = field { return number }
        if case .string(let text) = field { return Double(text) }
        return nil
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ContentToolPack1Logic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.offline.OfflineCacheKey
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Pure CT.M5 pack-1 decisions — attempt/reveal gating, commit immutability,
 * Class Pulse poll backoff, offline action queue ordering, client allowlist,
 * and Review/SRS cache invalidation keys. No networking.
 */
object ContentToolPack1Logic {
    val pack1ToolIds: Set<String> = setOf(
        "inline_questions",
        "predict_reveal",
        "class_pulse",
        "flashcards",
    )

    /** Per-tool client allowlist (rollout). Empty entry removes a renderer without a release. */
    var clientAllowlist: Set<String> = pack1ToolIds

    const val CLASS_PULSE_POLL_INTERVAL_MS = 30_000
    const val CLASS_PULSE_MAX_BACKOFF_MS = 120_000
    val flashcardRatings: List<String> = listOf("again", "hard", "good", "easy")

    data class PendingAction(
        val instanceId: String,
        val toolId: String,
        val action: String,
        val sequence: Long,
        val payloadJSON: String,
    )

    // MARK: - Allowlist / registry

    fun isClientAllowlisted(
        toolId: String,
        allowlist: Set<String> = clientAllowlist,
    ): Boolean = toolId in allowlist

    fun allowlistedToolIds(allowlist: Set<String> = clientAllowlist): Set<String> =
        pack1ToolIds.intersect(allowlist)

    fun conflictPolicy(toolId: String): ContentToolHostLogic.ConflictPolicy =
        if (toolId == "flashcards") {
            ContentToolHostLogic.ConflictPolicy.MERGE
        } else {
            ContentToolHostLogic.ConflictPolicy.SERVER_WINS
        }

    // MARK: - Offline action queue

    fun canQueueActionOffline(toolId: String, action: String): Boolean =
        (toolId == "inline_questions" && action == "submit") ||
            (toolId == "flashcards" && action == "rate")

    fun orderPendingActions(items: List<PendingAction>): List<PendingAction> =
        items.sortedWith(compareBy({ it.instanceId }, { it.sequence }))

    // MARK: - inline_questions

    /** null = unlimited */
    fun parseAttemptsConfig(raw: JsonElement?): Int? {
        if (raw == null) return 2
        val prim = raw as? JsonPrimitive ?: return 2
        val content = prim.contentOrNull ?: return 2
        if (content.equals("unlimited", ignoreCase = true)) return null
        return content.toIntOrNull()?.coerceAtLeast(1)
            ?: prim.doubleOrNull?.toInt()?.coerceAtLeast(1)
            ?: 2
    }

    fun attemptsUsed(answers: Map<String, JsonElement>, questionId: String): Int {
        val q = answers[questionId]?.asObjectMap() ?: return 0
        val attempts = q["attempts"] ?: return 0
        return runCatching { attempts.jsonArray.size }.getOrDefault(0)
    }

    fun canSubmit(
        answers: Map<String, JsonElement>,
        questionId: String,
        maxAttempts: Int?,
        readOnly: Boolean,
    ): Boolean {
        if (readOnly) return false
        if (maxAttempts == null) return true
        return attemptsUsed(answers, questionId) < maxAttempts
    }

    fun isSequentiallyUnlocked(
        questions: List<String>,
        answers: Map<String, JsonElement>,
        questionId: String,
        sequential: Boolean,
    ): Boolean {
        if (!sequential) return true
        for (qid in questions) {
            if (qid == questionId) return true
            if (attemptsUsed(answers, qid) == 0) return false
        }
        return false
    }

    /** How many questions to show at once. null means all. */
    fun parseQuestionsAtATime(raw: JsonElement?): Int? {
        if (raw == null) return null
        val prim = raw as? JsonPrimitive ?: return null
        val content = prim.contentOrNull ?: return null
        if (content.equals("all", ignoreCase = true)) return null
        val n = content.toIntOrNull() ?: prim.doubleOrNull?.toInt() ?: return null
        return if (n in 1..3) n else null
    }

    /** Inclusive-exclusive window [start, end) of the visible page. */
    fun pageWindow(total: Int, pageSize: Int?, pageIndex: Int): IntRange {
        if (pageSize == null || pageSize <= 0 || pageSize >= total || total <= 0) {
            return 0 until maxOf(0, total)
        }
        val pageCount = (total + pageSize - 1) / pageSize
        val page = pageIndex.coerceIn(0, pageCount - 1)
        val start = page * pageSize
        val end = minOf(total, start + pageSize)
        return start until end
    }

    fun initialPageIndex(total: Int, pageSize: Int?, firstIncompleteIndex: Int): Int {
        if (pageSize == null || pageSize <= 0 || total <= 0) return 0
        val idx = firstIncompleteIndex.coerceIn(0, total - 1)
        return idx / pageSize
    }

    fun <T> shuffleStable(items: List<T>, seed: String): List<T> {
        if (items.size <= 1) return items
        val out = items.toMutableList()
        var h = 0u
        for (u in seed.toByteArray()) {
            h = h * 31u + u.toUByte().toUInt()
        }
        for (i in out.lastIndex downTo 1) {
            h = h * 1_103_515_245u + 12_345u
            val j = (h % (i + 1).toUInt()).toInt()
            val tmp = out[i]
            out[i] = out[j]
            out[j] = tmp
        }
        return out
    }

    // MARK: - predict_reveal

    fun isCommitted(state: JsonElement?): Boolean {
        val committed = ContentToolHostLogic.stringField(state, "committedAt")?.trim().orEmpty()
        return committed.isNotEmpty()
    }

    fun canShowReveal(committed: Boolean, hasRevealPayload: Boolean): Boolean =
        committed && hasRevealPayload

    fun canEditPrediction(committed: Boolean, readOnly: Boolean): Boolean =
        !committed && !readOnly

    // MARK: - class_pulse

    fun hasVoted(votes: List<JsonElement>, round: Int = 1): Boolean =
        votes.any { vote ->
            val obj = vote.asObjectMap()
            val r = when (val raw = obj["round"]) {
                is JsonPrimitive -> raw.contentOrNull?.toIntOrNull() ?: raw.doubleOrNull?.toInt()
                else -> null
            }
            r == round
        }

    fun shouldPollAggregate(visible: Boolean, hasVoted: Boolean): Boolean =
        visible && hasVoted

    fun nextPollDelayMs(
        consecutiveFailures: Int,
        baseMs: Int = CLASS_PULSE_POLL_INTERVAL_MS,
    ): Int {
        if (consecutiveFailures <= 0) return baseMs
        val factor = 1 shl minOf(consecutiveFailures, 3)
        return minOf(CLASS_PULSE_MAX_BACKOFF_MS, baseMs * factor)
    }

    // MARK: - flashcards / Review reconciliation

    fun isValidRating(rating: String): Boolean =
        rating.lowercase() in flashcardRatings

    fun reviewCacheKeysToInvalidate(): List<String> =
        listOf(OfflineCacheKey.reviewQueue(), OfflineCacheKey.reviewStats())

    fun shouldDoubleCountReviewSubmit(toolId: String): Boolean =
        toolId != "flashcards"

    // MARK: - Unknown-field preservation

    fun mergePreservingUnknown(
        base: Map<String, JsonElement>,
        patch: Map<String, JsonElement>,
    ): Map<String, JsonElement> = base + patch

    fun objectMap(value: JsonElement?): Map<String, JsonElement> =
        value?.asObjectMap() ?: emptyMap()

    fun arrayField(value: JsonElement?, key: String): List<JsonElement> {
        val arr = value?.asObjectMap()?.get(key) ?: return emptyList()
        return runCatching { arr.jsonArray.toList() }.getOrDefault(emptyList())
    }

    fun boolField(value: JsonElement?, key: String): Boolean? {
        val prim = value?.asObjectMap()?.get(key) as? JsonPrimitive ?: return null
        return prim.booleanOrNull
    }

    fun numberField(value: JsonElement?, key: String): Double? {
        val prim = value?.asObjectMap()?.get(key) as? JsonPrimitive ?: return null
        return prim.doubleOrNull ?: prim.contentOrNull?.toDoubleOrNull()
    }

    private fun JsonElement.asObjectMap(): Map<String, JsonElement> =
        runCatching { jsonObject.toMap() }.getOrDefault(emptyMap())
}
